Backend/FcEvaluate.hs

```diff
@@ -112,30 +109,56 @@ matchTheAlts  dc  args ((FcAltFc (FcConPat dc' xs) rhs):rest)
           pure $ Just $ substFcTmInTm sub rhs
   | otherwise = matchTheAlts dc args rest
 
-
--- collapseProgram :: FcProgram 'Fc -> FcTerm 'Fc
--- smallStep :: MonadUnique m => FcTerm 'Fc -> m (Maybe (FcTerm 'Fc))
-
-groundTerm :: FcTerm 'Fc -> FcTerm 'Fc
-groundTerm (FcTmTyAbs ty t1) = FcTmTyApp (FcTmTyAbs ty (groundTerm t1)) FcTyUnit
-groundTerm t                 = t
-
--- | Follows lazy operations semantics
+-- | Perform a multi-step evaluation using lazy operational semantics. This
+-- | function performs single-step evaluations until the result of this single
+-- | step evaluation indicates that the given term is fully evaluated. If the
+-- | term is an error, then the error's string is returned, otherwise the
+-- | full term is returned.
 fullStep :: FcTerm 'Fc -> EvM (Either String (FcTerm 'Fc))
 fullStep t = smallStep t >>= \case
   Just t'  -> fullStep t'
   Nothing  -> case t of
     FcTmERROR err _ -> return $ Left err
     res             -> return $ Right res
 
--- | Fully evaluates all branches, by calling fullStep
+-- | Convert a program to a simple expression (local let-bindings).
+collapseProgram :: FcProgram 'Fc -> FcTerm 'Fc
+collapseProgram = \case
+  FcPgmTerm e                        ->  e
+  FcPgmDataDecl _ p                  -> collapseProgram p
+  FcPgmValDecl (FcValBind x ty e1) p -> FcTmLet x ty e1 $ collapseProgram p
+
+-- | Grounds the given term by wrapping all top level type abstractions in a
+-- | type application on unit. This is only patched grounding, and so no
+-- | internal type abstractions are grounded. For example:
+-- |   /\ a. let f = <expr1> in <expr2>
+-- | is grounded to:
+-- |   (/\ a. let f = <expr1> in <expr2>) Unit
+-- |
+-- | But (let f = <expr1> in /\a. <expr2>) is left unchanged. We do this
+-- | because we know the elaboration & desugaring phases only produce top level
+-- | type abstractions. So this method will be sufficient in all cases.
```

Comment:
```suggestion
-- | Grounds the given term by looking for top-level type abstractions, and
-- then applying it on as many unit values as needed. This is only patched
-- grounding, and so no internal type abstractions are grounded. For example:
--
-- > /\ a. let f = <expr1> in <expr2>
--
-- is grounded to:
--
-- > (/\ a. let f = <expr1> in <expr2>) Unit
-- 
-- But @(let f = <expr1> in /\a. <expr2>)@ is left unchanged. We do this
-- because we know the elaboration and desugaring phases only produce top-level
-- type abstractions. So this method will be sufficient in all cases.
-- TODO: In the future, this function should be type-directed: the type should
-- determine whether we should create an application to unit or not.
```
@reilem For haddock-compatible markup comments look here: https://www.haskell.org/haddock/doc/html/ch03s08.html